sympy/simplify/cse_main.py

```diff
@@ -212,42 +212,120 @@ def _find_opts(expr):
 
     ## Process Adds and commutative Muls
 
-    def _match_common_args(Func, funcs):
+    def _match_common_args(Func, funcs, dict, F):
         if order != 'none':
             funcs = list(ordered(funcs))
         else:
             funcs = sorted(funcs, key=lambda x: len(x.args))
 
-        func_args = [set(e.args) for e in funcs]
+        # ----------------- helpers ---------------------------
+        def cfunc(dx, dy, c):
+            # compute the common function from the common `c` keys
+            # SHARES Func
+            args = []
+            for a in c:
+                if not (
+                        dx[a].is_Integer and dx[a] > 0 and
+                        dy[a].is_Integer and dy[a] > 0):
+                    continue
+                b = min(dx[a], dy[a])
+                args.append(F(a, b))
+            return Func(*args)
+
+        def ufunc(*args):
+            # return a well formed unevaluated function from the args
+            # SHARES Func
+            from sympy.core.add import _addsort
+            from sympy.core.mul import _mulsort
+            args = list(args)
+            if Func is Add:
+                _addsort(args)
+            elif Func is Mul:
+                _mulsort(args)
+            else:
+                assert None # only Add or Mul expected
+            return Func(*args, evaluate=False)
+
+        def as_dict(e):
+            # return a well formed unevaluated function from the args
+            # SHARES dict, F
+            d = getattr(e, dict, lambda: {a: S.One for a in e.args})()
+            for k in list(d.keys()):
+                if not d[k].is_Integer or d[k] < 0:
+                    d[F(k, d.pop(k))] = S.One
+            return d
+
+        def from_dict(d):
+            # creates an expression from a coefficients or powers dict
+            # SHARES F
+            return ufunc(*[F(k, v) for k,v in d.items()])
+
+        def update(k):
+            # updates all of the info associated with k using
+            # the com_dict
+            # SHARES com_dict, func_dicts, func_args, opt_subs, funcs
+            verbose = False
+            for di in com_dict:
+                # don't allow a sign to change
+                if com_dict[di] > func_dicts[k][di]:
+                    return
+            # remove it
```

Comment:
remove it as many times as possible -- see test failure where `x+y` is extracted once from `4x+4y`